File: PA76.java
//Leetcode 720. Longest Word in Dictionary
//TTime complexity : O(nk) , n is the no of words and k is their average length.
//Space Complexity : O(nk) , n is the no of words and k is their average length.

class Solution {
       class TrieNode{
            TrieNode[] children;
            String word;
            TrieNode(){
                children = new TrieNode[26];                
            }
        }
     TrieNode root= new TrieNode();
      
    public void buildTrie(String word){
        TrieNode n= root;
        for(int i=0; i<word.length();i++){
            char c= word.charAt(i);
            if(n.children[c-'a']==null){
                
                n.children[c-'a'] = new TrieNode();
            }
            n = n.children[c-'a']; 
            
        }
        n.word= word;  
    }
    public String longestWord(String[] words) {
        
        if(words ==null || words.length ==0) return "";
        for(String w: words){
            buildTrie(w);
        }   
        Queue<TrieNode> q = new LinkedList<>();
        q.offer(root);
        TrieNode t=null;
        while(!q.isEmpty()){
              t= q.poll();
              for(int i=25;i>=0;i--){
                 if(t.children[i]!=null && t.children[i].word!=null){
                     q.offer(t.children[i]);
                 } 
              }         
        }
        
        return t.word;
        
    }
}

File: PA_76.py
#LC 720 - Longest Word in Dictionary
#Time Complexity - O(n.k) k being the average length of the word
#Space Complexity - O(n)
from collections import deque
class TrieNode(object):
    
    def __init__(self):
        self.children = dict()
        self.endOfWord = False
        self.word = ""
        
class Solution(object):
    def longestWord(self, words):
        """
        :type words: List[str]
        :rtype: str
        """
        self.root = TrieNode()
       
        def insert(word):
            """
            Inserts a word into the trie.
            :type word: str
            :rtype: None
            """
            currNode = self.root
            for i in range(len(word)):
                char = word[i]
                if char not in currNode.children:
                    currNode.children[char] = TrieNode()
                currNode = currNode.children[char]
            currNode.endOfWord = True
            currNode.word = word
        
        for i in words:
            insert(i)
        
        q = deque()
        q.append(self.root)
        ans = ""
        while q:
            node = q.popleft()
            if len(node.word) >= len(ans):
                if len(node.word) > len(ans) or ans > node.word:
                    ans = node.word
            for child in node.children:
                if node.children[child].endOfWord:
                    q.append(node.children[child])
        return(ans)